# Course in Advanced Bioinformatics

## Table of Contents

- [Course in Advanced Bioinformatics](#course-in-advanced-bioinformatics)
  - [Table of Contents](#table-of-contents)
- [Outline](#outline)
- [Shared point with students](#shared-point-with-students)
- [Software requirements](#software-requirements)
  - [Running with our Docker Image](#running-with-our-docker-image)
  - [Running locally](#running-locally)
- [Some EXTRA recommendations](#some-extra-recommendations)
- [Author](#author)
- [Contributing](#contributing)

# Outline

- Lecture 1: Introduction to genome sequencing and next generation sequencing methods.
- Lecture 2: Introduction to the Linux command line, NGS data formats, read mapping and alignments (hands-on available here).
- Lecture 3: NGS applications: RNA‐seq.
- Lecture 4: NGS applications: ChIP‐seq.
- Lecture 5: NGS applications: Variant detection, SNPs, CNVs, structural variants, epigenetic variation.
- Lecture 6: NGS applications: Sequence assembly, de‐novo sequencing and EST sequencing.
- Lecture 7: Introduction to Genome Browsing with the UCSC Genome Browser.

# Shared point with students

Lessons and schedule available [here](https://drive.google.com/drive/u/0/folders/1ezwEiPyYDk-OCIRkhAaS8J7yCK9aGBBo).

# Software requirements

## Running with our Docker Image

* [osvaldogc/ufv:2.0](https://hub.docker.com/r/osvaldogc/ufv/): Docker image with `Hands On` packages already installed.
Images can be pulled via `docker pull osvaldogc/ufv:2.0`. Detailed instructions [here](https://github.com/fpozoc/advanced-bioinformatics-course/-/tree/master/ADB-02-ngs-linux/handson#21-first-steps-with-docker).

## Running locally

Remember that all the packages required to run the analysis are already installed in our [Docker image](https://hub.docker.com/r/osvaldogc/ufv). 

Follow this [guide](https://github.com/fpozoc/advanced-bioinformatics-course/-/tree/master/ADB-02-ngs-linux/handson/local_installation) if you want to run your analysis locally.

# Some EXTRA recommendations

If you want to complement, add or step up your knowledge in bioinformatics... 

- My June 27, 2022 post in FEBS Network: [Why and how should you start learning bioinformatics?](https://network.febs.org/posts/why-and-how-should-you-start-learning-bioinformatics?channel_id=723-early-career-scientist)

Some deeper and longer... but highly recommended bioinformatics courses (still available online):

- [Foundations of Computational and Systems Biology](https://ocw.mit.edu/courses/biology/7-91j-foundations-of-computational-and-systems-biology-spring-2014/). MIT Opencourseware. Spring 2014.
- [MIT CompBio](https://www.youtube.com/watch?v=sX4cMu9Azgs&list=PLypiXJdtIca6GBQwDTo4bIEDV8F4RcAgt). MIT. 2018.

Recommended books:

- [Bioinformatics algorithms](https://www.bioinformaticsalgorithms.org/read-the-book) by Pavel A. Pevzner.
- [Modern Statistics for Modern Biology](https://www.huber.embl.de/msmb/) by Susan Holmes and Wolfgang Huber.

# Author

Fernando Pozo – [@fpozoca](https://twitter.com/fpozoca) - [Google Scholar](https://scholar.google.com/citations?user=3YLw4PQAAAAJ&hl=en) – fpozoc@cnio.es - [github.com/fpozoc](https://github.com/fpozoc)

# Contributing

1. Fork it (<https://github.com/fpozoc/advanced-bioinformatics-course>).
2. Create your feature branch (`git checkout -b feature/fooBar`).
3. Commit your changes (`git commit -am 'Add some fooBar'`).
4. Push to the branch (`git push origin feature/fooBar`).
5. Create a new Pull Request.
